File: tested/languages/java/templates/EvaluationResult.java
import java.util.ArrayList;
import java.util.List;

public class EvaluationResult {

    public final boolean result;
    public final String readableExpected;
    public final String readableActual;
    public final List<Message> messages;

    private EvaluationResult(boolean result, String readableExpected, String readableActual, List<Message> messages) {
        this.result = result;
        this.readableExpected = readableExpected;
        this.readableActual = readableActual;
        this.messages = messages;
    }

    public static Builder builder(boolean accepted) {
        return new Builder(accepted);
    }

    public static class Message {
        public final String description;
        public final String format;
        public final String permission;

        public Message(String description, String format, String permission) {
            this.description = description;
            this.format = format;
            this.permission = permission;
        }

        public Message(String description, String format) {
            this(description, format, null);
        }

        public Message(String description) {
            this(description, "text");
        }
    }

    public static class Builder {

        private final boolean result;
        private String readableExpected = null;
        private String readableActual = null;
        private final List<Message> messages = new ArrayList<>();

        private Builder(boolean result) {
            this.result = result;
        }

        public Builder withReadableExpected(String readableExpected) {
            this.readableExpected = readableExpected;
            return this;
        }

        public Builder withReadableActual(String readableActual) {
            this.readableActual = readableActual;
            return this;
        }

        public Builder withMessages(List<Message> messages) {
            this.messages.addAll(messages);
            return this;
        }

        public Builder withMessage(Message message) {
            this.messages.add(message);
            return this;
        }

        public EvaluationResult build() {
            return new EvaluationResult(result, readableExpected, readableActual, messages);
        }
    }
}

File: tested/languages/kotlin/templates/EvaluationResult.kt
class EvaluationResult private constructor(
        val result: Boolean,
        val readableExpected: String?,
        val readableActual: String?,
        val dslExpected: String?,
        val dslActual: String?,
        val messages: List<Message>
) {
    class Message(val description: String,
                  val format: String = "text",
                  val permission: String? = null)

    class Builder(
            private val result: Boolean,
            private var readableActual: String? = null,
            private var readableExpected: String? = null,
            private var dslActual: String? = null,
            private var dslExpected: String? = null
    ) {
        private val messages: MutableList<Message> = ArrayList()

        fun withReadableActual(readableActual: String?): Builder {
            this.readableActual = readableActual
            return this
        }

        fun withReadableExpected(readableExpected: String?): Builder {
            this.readableExpected = readableExpected
            return this
        }

        fun withDslActual(dslActual: String?): Builder {
            this.dslActual = dslActual
            return this
        }

        fun withDslExpected(dslExpected: String?): Builder {
            this.dslExpected = dslExpected
            return this
        }

        fun withMessages(messages: List<Message>): Builder {
            this.messages += messages
            return this
        }

        fun withMessage(message: Message): Builder {
            this.messages += message
            return this
        }

        fun build(): EvaluationResult {
            return EvaluationResult(result, readableExpected,
                    readableActual, dslExpected, dslActual, messages)
        }
    }
}
